### PR TITLE
Reduce allocations of scene items in the sw renderer

### PR DESCRIPTION
**Pull request by tronical**
Allocate capacity for scene item vector with the number of items from
the last scene.

That shaves off about 1-2ms of pre-scene time on the pico and stm32.

I'm not entirely sure that this is worth the complexity. Or perhaps there's a much simpler way?

### CODE REVIEW

**Comment by ogoffart** [comment]
Why not keep the state in the LineRenderer itself?   (yeah, this will be useless when using the rendering by buffer, but it's not so much data)

Also this is only for one of the vector, but there is a bunch of vector where we append items, in particular the texture vector should have about the same size. so this may be another ~2ms if we do it for the other vectors.

**Comment by tronical** [comment]
Sure, good idea - I don't mind keeping it there then. That's certainly less intrusive!

**Comment by tronical** [comment]
> Why not keep the state in the LineRenderer itself? (yeah, this will be useless when using the rendering by buffer, but it's not so much data)
> 
> Also this is only for one of the vector, but there is a bunch of vector where we append items, in particular the texture vector should have about the same size. so this may be another ~2ms if we do it for the other vectors.

Done. This is less intrusive now.

**Comment by ogoffart** [comment]
How is the speed with that path now?  Does it improve perfs on the pico?

File: internal/core/swrenderer.rs [hunk 1]
```diff
@@ -524,13 +526,39 @@ impl<'a, T: TargetPixel> ProcessScene for RenderToBuffer<'a, T> {
...
 
-#[derive(Default)]
+#[derive(Copy, Clone, Default)]
+struct SceneCapacities {
+    num_items: usize,
+    num_textures: usize,
+    num_rounded_rectangles: usize,
+}
+
+impl From<&Scene> for SceneCapacities {
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/core/swrenderer.rs:536]: Hehe sure, I'll change it :)

**Comment by tronical** [comment]
> How is the speed with that path now? Does it improve perfs on the pico?

My pico is in the office :(. I can check next week.

**Comment by tronical** [comment]
Deferring this for now as I'm not working on this at the moment.

**Pull request closed without merging**